src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php

```diff
@@ -1260,38 +1260,75 @@ protected function prepare_item_for_database( $request ) {
 	 * @param WP_Post_Type $post_type   Post type.
 	 * @return string|WP_Error Post status or WP_Error if lacking the proper permission.
 	 */
-	protected function handle_status_param( $post_status, $post_type ) {
-
-		switch ( $post_status ) {
-			case 'draft':
-			case 'pending':
-				break;
-			case 'private':
-				if ( ! current_user_can( $post_type->cap->publish_posts ) ) {
-					return new WP_Error(
-						'rest_cannot_publish',
-						__( 'Sorry, you are not allowed to create private posts in this post type.' ),
-						array( 'status' => rest_authorization_required_code() )
-					);
-				}
-				break;
-			case 'publish':
-			case 'future':
-				if ( ! current_user_can( $post_type->cap->publish_posts ) ) {
-					return new WP_Error(
-						'rest_cannot_publish',
-						__( 'Sorry, you are not allowed to publish posts in this post type.' ),
-						array( 'status' => rest_authorization_required_code() )
-					);
-				}
-				break;
-			default:
-				if ( ! get_post_status_object( $post_status ) ) {
-					$post_status = 'draft';
-				}
-				break;
+	protected function handle_status_param( $post_status, $post_type, $post_id = null ) {
+
+		if( $post_id ) {
+			$existing_post = $this->get_post( $post_id );
+
+			switch ( $post_status ) {
+				case 'draft':
+				case 'pending':
+				case 'private':
+					if ( ! current_user_can( 'edit_post', $post_id ) ) {
+						return new WP_Error(
+							'rest_cannot_publish',
+							__( 'Sorry, you are not allowed to edit this post.' ),
+							array( 'status' => rest_authorization_required_code() )
+						);
+					}
+					break;
+				case 'publish':
+				case 'future':
+					if (
+						! ( in_array($existing_post->post_status, array( 'publish', 'future' ) ) && current_user_can( 'edit_post', $post_id ) ) 
```

Comment:
```suggestion
						! ( in_array( $existing_post->post_status, array( 'publish', 'future' ), true ) && current_user_can( 'edit_post', $post_id ) ) 
```

WordPress requires a space after parentheses, and we enforce using the strict mode for `in_array`.